File: design-patterns/template-method/report.rb
class Report
  attr_reader :title, :text

  def initialize
    @title = 'Monthly Report'
    @text = ['Things are going', 'really, really well.']
  end
  
  def output_report
    output_start
    output_head
    output_body_start
    output_body
    output_body_end
    output_end
  end
  
  def output_body
    text.each do |line|
      output_line(line)
    end
  end
  
  def output_start
    raise "Called abstract method: #{__method__}"
  end
  
  def output_head
    raise "Called abstract method: #{__method__}"
  end
  
  def output_body_start
    raise "Called abstract method: #{__method__}"
  end
  
  def output_line(line)
    raise "Called abstract method: #{__method__}"
  end
  
  def output_body_end
    raise "Called abstract method: #{__method__}"
  end
  
  def output_end
    raise "Called abstract method: #{__method__}"
  end
end

File: design-patterns/java/DesignPatterns/src/ar/edu/unlam/cdp/patterns/templateMethod/Report.java
package ar.edu.unlam.cdp.patterns.templateMethod;

import java.util.LinkedList;
import java.util.List;

public abstract class Report {

	private String title;
	private List<String> text = new LinkedList<String>();
	
	public Report() {
		this.title = "Monthly Report";
		this.text.add("Things are going");
		this.text.add("really, really well.");
	}
	
	public final void outputReport(){
	    this.outputStart();
	    this.outputHead();
	    this.outputBodyStart();
	    this.outputBody();
	    this.outputBodyEnd();
	    this.outputEnd();
	}
	
	public void outputBody() {
		for (String line : this.text) {
			this.outputLine(line);
		}
	}
	
	public abstract void outputLine(String line);

	public abstract void outputStart();
	public abstract void outputHead();
	public abstract void outputBodyStart();
	public abstract void outputBodyEnd();
	public abstract void outputEnd();

	public String getTitle() {
		return title;
	}

	public List<String> getText() {
		return text;
	}

}
